File: src/com/leozhi/topic125/Solution03.java
package com.leozhi.topic125;

/**
 * @author leozhi
 * 筛选 + 左右指针
 * 通过
 * 5ms
 */
public class Solution03 {
    public boolean isPalindrome(String s) {
        StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (Character.isLetterOrDigit(s.charAt(i))) {
                stringBuilder.append(s.charAt(i));
            }
        }
        s = stringBuilder.toString().toLowerCase();
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) == s.charAt(right)) {
                left++;
                right--;
            } else {
                return false;
            }
        }
        return true;
    }
}

File: src/com/leozhi/topic201/Solution03.kt
package com.leozhi.topic201

/**
 * @author leozhi
 * Brian Kernighan 算法
 * 官方题解
 * 通过
 * 260ms
 */
class Solution03 {
    fun rangeBitwiseAnd(left: Int, right: Int): Int {
        var r = right
        while (left < r) {
            r = r.and(r - 1)
        }
        return r
    }
}